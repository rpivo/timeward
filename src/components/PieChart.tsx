import React, { useContext } from 'react';
import { DashboardContext } from '@pages/Dashboard';
import StyledPieChart from '@styles/components/PieChart.styled';

const PieChart: React.FC = (): JSX.Element => {
  const { store, totalSeconds } = useContext(DashboardContext);
  const strokes: number[] = [];

  const getPreviousStrokeLengths = (index: number): number =>
    strokes
      .slice(0, index)
      .reduce((acc, curr) => acc + curr, 0);

  const getRotation = (index: number): number =>
    ((getPreviousStrokeLengths(index) / 157) * 360) - 90;

  const getRandomHexColor = (): string => `#${(Math.random() * 0xFFFFFF << 0).toString(16)}`;

  return (
    <StyledPieChart>
      <svg
        width='100'
        height='100'
        viewBox='0 0 100 100'
      >
      </svg>
      {store.map((record, index) => {
        const stroke = ~~((record.seconds / totalSeconds) * 157);
        strokes.push(stroke);
        if (record.seconds) return (
          <svg
            key={index}
            width='100'
            height='100'
            viewBox='0 0 100 100'
            style={{ transform: `rotate(${getRotation(index)}deg)` }}
          >
            <circle
              r='25'
              cx='50'
              cy='50'
              fill='transparent'
              stroke={getRandomHexColor()}
              strokeDasharray={`${stroke} 157`}
              strokeWidth='50'
            />
          </svg>
        );
        return null;
      })}
    </StyledPieChart>
  );
};

export default PieChart;

