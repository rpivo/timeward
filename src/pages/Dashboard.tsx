import React, { createContext, useReducer, useRef } from 'react';
import Alignment from '@components/Alignment';
import Graph from '@components/Graph';
import PieChart from '@components/PieChart';
import Section from '@components/Section';
import Tile from '@components/Tile';
import Timeline from '@components/Timeline';
import Timer from '@components/Timer';
import Timesheet from '@components/Timesheet';
import StyledDashboard from '@styles/pages/Dashboard.styled';

type Action = {
  type: 'button clicked';
  payload: number;
};
type Store = {
  label: string | undefined;
  seconds: number;
}[];

type DashboardContextType = {
  dispatch: React.Dispatch<Action>;
  store: Store;
}

export const DashboardContext = createContext({} as DashboardContextType);

const Dashboard: React.FC = (): JSX.Element => {

  const ref = useRef<HTMLInputElement>(null);

  const reducer = (state: Store, action: Action): Store => {
    switch (action.type) {
    case 'button clicked':
      return [
        ...state,
        {
          label: ref?.current?.value,
          seconds: action.payload,
        },
      ];
    default:
      return state;
    }
  };

  const dashboardStore = [
    {
      label: '',
      seconds: 0,
    },
  ];
  const [store, dispatch] = useReducer(reducer, dashboardStore);

  const handleBlur = (): void => console.log(ref?.current?.value);

  return (
    <StyledDashboard>
      <Section>
        <DashboardContext.Provider value={{ dispatch, store }}>
          <Tile>
            <Alignment horizontal>
              <input ref={ref} type='text' onBlur={(): void => handleBlur()}></input>
            </Alignment>
            <Alignment horizontal>
              <Timer />
            </Alignment>
            <Timesheet>
              <Timesheet.Record />
            </Timesheet>
          </Tile>
        </DashboardContext.Provider>
        <Tile>
          <Graph />
        </Tile>
      </Section>
      <PieChart />
      <Tile width='full'>
        <Timeline>
          <Timeline.Day day={0} />
          <Timeline.Day day={1} />
          <Timeline.Day day={2} />
          <Timeline.Day day={3} />
          <Timeline.Day day={4} />
          <Timeline.Day day={5} />
          <Timeline.Day day={6} />
        </Timeline>
      </Tile>
    </StyledDashboard >
  );
};

export default Dashboard;
