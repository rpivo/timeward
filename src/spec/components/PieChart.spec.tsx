import React from 'react';
import renderer from 'react-test-renderer';
import { DashboardContext } from '@pages/Dashboard';
import PieChart from '@components/PieChart';

describe('PieChart', () => {
  it('should render correctly', () => {
    const dispatch = jest.fn();
    const store = [
      {
        color: '#000000',
        label: 'work',
        seconds: 10,
      },
      {
        color: '#111111',
        label: 'play',
        seconds: 20,
      },
    ];
    const totalSeconds = 0;
    const tree = renderer
      .create(
        <DashboardContext.Provider value={{ dispatch, store, totalSeconds }}>
          <PieChart />
        </DashboardContext.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
