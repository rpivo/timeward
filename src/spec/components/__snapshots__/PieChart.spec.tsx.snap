// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart should render correctly 1`] = `
<div
  className="sc-AxiKw hIppFK"
>
  <svg
    height="100"
    viewBox="0 0 100 100"
    width="100"
  />
  <svg
    height="100"
    style={
      Object {
        "transform": "rotate(-90deg)",
      }
    }
    viewBox="0 0 100 100"
    width="100"
  >
    <circle
      cx="50"
      cy="50"
      fill="transparent"
      r="25"
      stroke="#000000"
      strokeDasharray="0 157"
      strokeWidth="50"
    />
  </svg>
  <svg
    height="100"
    style={
      Object {
        "transform": "rotate(-90deg)",
      }
    }
    viewBox="0 0 100 100"
    width="100"
  >
    <circle
      cx="50"
      cy="50"
      fill="transparent"
      r="25"
      stroke="#111111"
      strokeDasharray="0 157"
      strokeWidth="50"
    />
  </svg>
</div>
`;
